ui: add an Overload dashboard

RaduBerinde opened the pull request:
This commit adds an Overload dashboard in the metrics view. This is
intended to be a convenient way to monitor admission control.

The dashboard contains:
 - SQL Queries
 - Service latency
 - CPU Percent
 - Runnable Goroutines per CPU
 - L0 Sublevels and Files

Release note (ui change): a new Overload dashboard groups metrics that
are useful for admission control.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/66595)
<!-- Reviewable:end -->

RaduBerinde commented:
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-McPUiYLEk8lqv083dub:-McPUiYLEk8lqv083duc:b-br97hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133832671a0d291dfb205deddc3d8ad9c628842/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L128)):*
> ```XML
>     >
>       <Axis label="count">
>         {nodeIDs.map((nid) => (
> ```

I couldn't figure out how to group these so I can show both metrics for n1, then both metrics for n2, etc. I tried putting both `<Metric>`s in the same  `=> ( )`, not sure what I'm doing wrong.


<!-- Sent from Reviewable.io -->

dhartunian reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @RaduBerinde and @sumeerbhola)

___
*[pkg/ts/catalog/chart_catalog.go, line 2271 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-Mcixh163bImrO8Zu-FX:-Mcixh163bImrO8Zu-FY:b-h3ttul) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133832671a0d291dfb205deddc3d8ad9c628842/pkg/ts/catalog/chart_catalog.go#L2271)):*
> ```Go
> 					"rocksdb.compacted-bytes-read",
> 					"rocksdb.compacted-bytes-written",
> 					"rocksdb.flushed-bytes",
> ```

Minor: why is this repeated here?
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-McPUiYLEk8lqv083dub:-McivtLrC5xC3Tiq2T0W:b-uk8lsk) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133832671a0d291dfb205deddc3d8ad9c628842/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L128)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I couldn't figure out how to group these so I can show both metrics for n1, then both metrics for n2, etc. I tried putting both `<Metric>`s in the same  `=> ( )`, not sure what I'm doing wrong.
</blockquote></details>

You might have been missing one of: `{}` instead of `()`, `return` in the lambda, or `<>` to group the multiple `<Metrics>` into a single JSX return object. If your lambda looks like the following it should work:

```
(nid) => {
  ...
  return (
    <>
    <Metric ... />
    <Metric ... />
    </>
  ) 
}
```

If it doesn't maybe try flatmap. I think the code that renders the plots, iterates through the children of `<Axis>` so it might get confused if they're grouped in 2s:

nodeIDs.flatMap((nid) => {
  ...
  return [
    <Metric ... />,
    <Metric ... />
  ] 
})
```
```



<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @sumeerbhola)

___
*[pkg/ts/catalog/chart_catalog.go, line 2271 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-Mcixh163bImrO8Zu-FX:-McoJd1h3KSDuYeRkCHI:bipe3gm) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133832671a0d291dfb205deddc3d8ad9c628842/pkg/ts/catalog/chart_catalog.go#L2271)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

Minor: why is this repeated here?
</blockquote></details>

Oops, no idea. Fixed.
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 128 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-McPUiYLEk8lqv083dub:-McoJf059-QzOMKnUAj8:bdmdlri) ([raw file](https://github.com/cockroachdb/cockroach/blob/7133832671a0d291dfb205deddc3d8ad9c628842/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L128)):*
<details><summary><i>Previously, dhartunian (David Hartunian) wrote…</i></summary><blockquote>

You might have been missing one of: `{}` instead of `()`, `return` in the lambda, or `<>` to group the multiple `<Metrics>` into a single JSX return object. If your lambda looks like the following it should work:

```
(nid) => {
  ...
  return (
    <>
    <Metric ... />
    <Metric ... />
    </>
  ) 
}
```

If it doesn't maybe try flatmap. I think the code that renders the plots, iterates through the children of `<Axis>` so it might get confused if they're grouped in 2s:

nodeIDs.flatMap((nid) => {
  ...
  return [
    <Metric ... />,
    <Metric ... />
  ] 
})
```
</blockquote></details>

Your `<>` sorcery worked! Thanks!


<!-- Sent from Reviewable.io -->

RaduBerinde commented:
@sumeerbhola let me know if this is ok with you. This is not necessarily what I'm proposing we'll have in the release, but at least for now having all these on the same page would be useful to understand what admission control is doing.

sumeerbhola commented:
> @sumeerbhola let me know if this is ok with you

I meant to look at this on monday but am very behind on things because of oncall.

sumeerbhola reviewed (changes requested):
sorry about the delay

> Reviewed 3 of 5 files at r1, 2 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-:-MctzIy4EOkwwT3JsnCe:b563jq1)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian and @RaduBerinde)

___
*[pkg/kv/kvserver/metrics.go, line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctxN4h1aH-lk-B4y1f:-MctwAB2Ak9r0uAQd3mn:b-e0jb1d) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/kv/kvserver/metrics.go#L410)):*
> ```Go
> 	}
> 	metaRdbL0Sublevels = metric.Metadata{
> 		Name:        "rocksdb.l0-sublevels",
> ```

There is no concept of sub-levels in rocksdb. I think we should start adding new metrics like this as pebble.l0-sublevels even though it potentially hurts discoverability.
@petermattis @itsbilal @jbowens do you have an opinion?
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctyDmT2HBObM61Tbp4:-MctyDmT2HBObM61Tbp5:b-w4xvmt) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L36)):*
> ```XML
>   } = props;
> 
>   return [
> ```

I am wondering whether sql latency should be on this dashboard. I think of an overload dashboard as something that one goes to to see if there is a resource overload, after one has seen some bad externally visible behavior (like high sql latency) on other dashboards. So it would focus on "resources": cpu, LSM health, root-sql BytesMonitor capacity and usage, (and when we have it) contention metrics for latches and locks.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @RaduBerinde, and @sumeerbhola)

___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctyDmT2HBObM61Tbp4:-Mcu6N-xDiDXdL--muPL:b-81f7j2) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I am wondering whether sql latency should be on this dashboard. I think of an overload dashboard as something that one goes to to see if there is a resource overload, after one has seen some bad externally visible behavior (like high sql latency) on other dashboards. So it would focus on "resources": cpu, LSM health, root-sql BytesMonitor capacity and usage, (and when we have it) contention metrics for latches and locks.
</blockquote></details>

I agree with you in principle but I wanted to put them here for practical reasons - many times you are looking for correlations between the latency/throughput graphs and overload-related stats. It's hard to look for that if you have to keep switching between dashboards. It's even harder to extract a screenshot that shows a correlation.

If we had a way to see two dashboards one below the other, this wouldn't be necessary. Maybe I should file an issue about that?


<!-- Sent from Reviewable.io -->

itsbilal commented:
___
*[pkg/kv/kvserver/metrics.go, line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctxN4h1aH-lk-B4y1f:-Mcu8UiZ3FMH086I703y:b-3vrn3e) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/kv/kvserver/metrics.go#L410)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

There is no concept of sub-levels in rocksdb. I think we should start adding new metrics like this as pebble.l0-sublevels even though it potentially hurts discoverability.
@petermattis @itsbilal @jbowens do you have an opinion?
</blockquote></details>

Yeah I agree, we shouldn't be naming more things `rocksdb.` at least. I'd also prefer calling it `storage.l0-sublevels` as I'm not sure how discoverable the `pebble` name is, but either of those two is better than `rocksdb`.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jbowens, @petermattis, and @RaduBerinde)

___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctyDmT2HBObM61Tbp4:-McuFAnBBB3CMh1sVKAl:bisr50y) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L36)):*
> Maybe I should file an issue about that?

Yes, I would prefer that to be solved in a different manner. This issue occurs already, and I don't think we should pollute our curated dashboards to attempt to solve it.
One approach I've seen is a pin beside each graph which one can click. All the pinned graphs can then be seen on a separate pinned dashboard.


<!-- Sent from Reviewable.io -->

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jbowens, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/metrics.go, line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctxN4h1aH-lk-B4y1f:-McuhR4Z4JtmSej1efbd:b-ycnkfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/kv/kvserver/metrics.go#L410)):*
<details><summary><i>Previously, itsbilal (Bilal Akhtar) wrote…</i></summary><blockquote>

Yeah I agree, we shouldn't be naming more things `rocksdb.` at least. I'd also prefer calling it `storage.l0-sublevels` as I'm not sure how discoverable the `pebble` name is, but either of those two is better than `rocksdb`.
</blockquote></details>

What would be the consequences of renaming all existing `rocksdb.*` to `pebble.` or 'storage.'?

In any case, seems like I'd be setting a precedent here so I want to make sure I make the right choice for these new stats.
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctyDmT2HBObM61Tbp4:-McuhuaJABh6soczejWK:bbzy92i) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L36)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

> Maybe I should file an issue about that?

Yes, I would prefer that to be solved in a different manner. This issue occurs already, and I don't think we should pollute our curated dashboards to attempt to solve it.
One approach I've seen is a pin beside each graph which one can click. All the pinned graphs can then be seen on a separate pinned dashboard.

</blockquote></details>

Filed https://github.com/cockroachdb/cockroach/issues/66798 and I'll remove the `SQL Queries` and `SQL Latency` graphs from this dashboard.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jbowens, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/metrics.go, line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctxN4h1aH-lk-B4y1f:-McyXhKW7pyxnC-OfO9c:b-u1yb60) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/kv/kvserver/metrics.go#L410)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

What would be the consequences of renaming all existing `rocksdb.*` to `pebble.` or 'storage.'?

In any case, seems like I'd be setting a precedent here so I want to make sure I make the right choice for these new stats.
</blockquote></details>

I don't think we want to rename existing ones. Customers may have tooling or custom dashboard links that rely on the existing names. So we'll have a mix of prefixes: rocksdb and pebble/storage.


<!-- Sent from Reviewable.io -->

RaduBerinde committed:
ui: add an Overload dashboard

This commit adds an Overload dashboard in the metrics view. This is
intended to be a convenient way to monitor admission control.

The dashboard contains:
 - CPU Percent
 - Runnable Goroutines per CPU
 - L0 Sublevels and Files

Release note (ui change): a new Overload dashboard groups metrics that
are useful for admission control.

RaduBerinde reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @jbowens, @petermattis, and @sumeerbhola)

___
*[pkg/kv/kvserver/metrics.go, line 410 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctxN4h1aH-lk-B4y1f:-MdJ8sJg0w-q5D0TukI3:b-yknit3) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/kv/kvserver/metrics.go#L410)):*
<details><summary><i>Previously, sumeerbhola wrote…</i></summary><blockquote>

I don't think we want to rename existing ones. Customers may have tooling or custom dashboard links that rely on the existing names. So we'll have a mix of prefixes: rocksdb and pebble/storage.
</blockquote></details>

Renamed the new ones to `storage.*`
___
*[pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx, line 36 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-MctyDmT2HBObM61Tbp4:-MdJ8wSeDuxPI1BL4pul:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/ed4b23560bf95d92e5854e2dcbf92b53dd65cdbd/pkg/ui/src/views/cluster/containers/nodeGraphs/dashboards/overload.tsx#L36)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Filed https://github.com/cockroachdb/cockroach/issues/66798 and I'll remove the `SQL Queries` and `SQL Latency` graphs from this dashboard.
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

sumeerbhola reviewed (approved):
<img class="emoji" title=":lgtm_strong:" alt=":lgtm_strong:" align="absmiddle" src="https://reviewable.io/lgtm_strong.png" height="20" width="61"/> 

> Reviewed 4 of 4 files at r3.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/66595#-:-MdJBj7X16cIZxQ4bLl9:b-uwkowd)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @dhartunian, @jbowens, @petermattis, and @sumeerbhola)



<!-- Sent from Reviewable.io -->

RaduBerinde commented:
bors r+

RaduBerinde commented:
Some acceptance flake.

bors r+
